dpsim-models/src/EMT/EMT_Ph3_SSN_Capacitor.cpp

```diff
@@ -0,0 +1,162 @@
+ *********************************************************************************/
+
+#include <dpsim-models/EMT/EMT_Ph3_SSN_Capacitor.h>
+
+using namespace CPS;
+
+EMT::Ph3::SSN::Capacitor::Capacitor(String uid, String name, Logger::Level logLevel)
+	: MNASimPowerComp<Real>(uid, name, true, true, logLevel), Base::Ph3::Capacitor(mAttributes) {
+	mPhaseType = PhaseType::ABC;
+    setVirtualNodeNumber(1);
+	setTerminalNumber(2);
+	mHistoricVoltage = Matrix::Zero(3, 1);	
+	**mIntfVoltage = Matrix::Zero(3, 1);
+	**mIntfCurrent = Matrix::Zero(3, 1);
+}
+
+
+SimPowerComp<Real>::Ptr EMT::Ph3::SSN::Capacitor::clone(String name) {
+	auto copy = Capacitor::make(name, mLogLevel);
+	copy->setParameters(**mCapacitance);
+	return copy;
+}
+void EMT::Ph3::SSN::Capacitor::initializeFromNodesAndTerminals(Real frequency) {
+
+	Real omega = 2 * PI * frequency;
+	MatrixComp admittance = MatrixComp::Zero(3, 3);
+	admittance <<
+		Complex(0, omega * (**mCapacitance)(0, 0)), Complex(0, omega * (**mCapacitance)(0, 1)), Complex(0, omega * (**mCapacitance)(0, 2)),
+		Complex(0, omega * (**mCapacitance)(1, 0)), Complex(0, omega * (**mCapacitance)(1, 1)), Complex(0, omega * (**mCapacitance)(1, 2)),
+		Complex(0, omega * (**mCapacitance)(2, 0)), Complex(0, omega * (**mCapacitance)(2, 1)), Complex(0, omega * (**mCapacitance)(2, 2));
+
+	MatrixComp vInitABC = Matrix::Zero(3, 1);
+	vInitABC(0, 0) = RMS3PH_TO_PEAK1PH * initialSingleVoltage(1) - RMS3PH_TO_PEAK1PH * initialSingleVoltage(0);
+	vInitABC(1, 0) = vInitABC(0, 0) * SHIFT_TO_PHASE_B;
+	vInitABC(2, 0) = vInitABC(0, 0) * SHIFT_TO_PHASE_C;
+	**mIntfVoltage = vInitABC.real();
+	**mIntfCurrent = (admittance * vInitABC).real();
+
+	SPDLOG_LOGGER_INFO(mSLog, "\nCapacitance [F]: {:s}"
+				"\nAdmittance [S]: {:s}",
+				Logger::matrixToString(**mCapacitance),
+				Logger::matrixCompToString(admittance));
+	SPDLOG_LOGGER_INFO(mSLog,
+		"\n--- Initialization from powerflow ---"
+		"\nVoltage across: {:s}"
+		"\nCurrent: {:s}"
+		"\nTerminal 0 voltage: {:s}"
+		"\nTerminal 1 voltage: {:s}"
+		"\n--- Initialization from powerflow finished ---",
+		Logger::matrixToString(**mIntfVoltage),
+		Logger::matrixToString(**mIntfCurrent),
+		Logger::phasorToString(RMS3PH_TO_PEAK1PH * initialSingleVoltage(0)),
+		Logger::phasorToString(RMS3PH_TO_PEAK1PH * initialSingleVoltage(1)));
+}
+
+void EMT::Ph3::SSN::Capacitor::mnaCompInitialize(Real omega, Real timeStep, Attribute<Matrix>::Ptr leftVector) {
+	updateMatrixNodeIndices();
+
+	mHistoricVoltage = Dufour_B_k_hat * **mIntfCurrent  + **mIntfVoltage;
+	Dufour_B_k_hat = timeStep * (2.0 * **mCapacitance).inverse();
+	Dufour_W_k_n = Dufour_B_k_hat;
+
+	**mRightVector = Matrix::Zero(leftVector->get().rows(), 1);
+}
+
+void EMT::Ph3::SSN::Capacitor::mnaCompApplySystemMatrixStamp(SparseMatrixRow& systemMatrix) {
+	if (terminalNotGrounded(0)) {
+		Math::addToMatrixElement(systemMatrix, matrixNodeIndex(0, 0), mVirtualNodes[0]->matrixNodeIndex(PhaseType::A), -1);
```

Comment:
In terms of Dufour's paper terminology: is `SSN_Inductor` a V_Type group, `SSN_Capacitor` an I_Type group and  `SSN_Full_Serial_RLC` a Mixed type group? I've noticed that the logic in `mnaCompApplySystemMatrixStamp()` for `SSN_Capacitor` is different from logic in `SSN_Inductor` and `SSN_Full_Serial_RLC.` Would it be possible to refactor the `SSN_Capacitor` code to have the same stamping logic, as you have managed to do it for `SSN_Full_Serial_RLC?`

Reply:
SSN_Inductor is a V_Type group because the output equation uses the unknown voltage to define the output current. The SSN_Capacitor is an I_Type group because it is the other way around; the unknown quantity which ends up in the left side vector sets the type.

For V_Type goups, the current is defined and the voltage is unknown; thus it has the exact same structure as the MNA approach, GV = I and can be directly stamped. For I_type groups we have the same stamping structure as for voltage sources, where the matrix is expanded by 1x1 (or for three phases 3x3). The extra line contains the compnent's output equation which adds a voltage to the excitation vector (rather than a current) and a current to the left side vector of unknowns (rather than a voltage); it is not a nodal but a mesh equation, so the system gets an extra "virtual node" to include the defining equation.

Following that logic, the SSN_Full_Serial_RLC is an I_Type group, not a mixed type group; if we had two output equations for example, with one output quantity being a current and one being a voltage, then we would have a mixed type group. The output equation for the RLC circuit defines an output current; we just have two state equations for two internal states.

In the Dufour paper it is stated that you can convert the group types into each other, but it would take extra steps and as we follow the MNA approach and not just the nodal approach without mesh equations, I did not look at how that would work, unfortunately.

Reply:
Thank you for a very nice and clear explanation of different types of groups, and for how you are integrating mesh equations into MNA. I am more of a Nodal Analysis guy myself😆, so for me that's a learning as well. This is pretty cool and powerful. Great work😉.
It should be quite easy to convert Thevenin Equivalent to Norton Equivalent, but not necessary though.